File: C#/Boggle/DiceTray.cs
using System;
using System.Collections.Generic;
using System.Text;

class DiceTray
{
    private static Random random = new Random();
    public Die[,] dice { get; private set; }
    private int compareLength { get; set; } 

    public DiceTray()
    {
        List<string> godsDice = new List<string>();
        godsDice.Add("LRYTTE");
        godsDice.Add("ANAEEG");
        godsDice.Add("AFPKFS");
        godsDice.Add("YLDEVR");
        godsDice.Add("VTHRWE");
        godsDice.Add("IDSYTT");
        godsDice.Add("XLDERI");
        godsDice.Add("ZNRNHL");
        godsDice.Add("EGHWNE");
        godsDice.Add("OATTOW");
        godsDice.Add("HCPOAS");
        godsDice.Add("NMIQHU");
        godsDice.Add("SEOTIS");
        godsDice.Add("MTOICU");
        godsDice.Add("ENSIEU");
        godsDice.Add("OBBAOJ");

        Shuffle<string>(godsDice);

        // Assign game pieces, and delete temporary dice
        this.dice = new Die[4,4];
        for (int row = 0; row < 4; row++)
        {
            for (int col = 0; col < 4; col++)
            {
                this.dice[row,col] = new Die(godsDice[0][random.Next(6)]);
                godsDice.RemoveAt(0);
            }
        }
    }

    public bool stringFound(string search)
    {
        search = search.ToUpper();
        this.compareLength = 0;
        char firstLetterOfSearch = search[0];

        for (int i = 0; i < 4; i++)
        {
            for (int j = 0; j < 4; j++)
            {
                if (firstLetterOfSearch == this.dice[i,j].letter)
                    if (this.stringFound(search, i, j))
                        return true;
            }
        }

        return false;
    }

    private bool stringFound(string search, int row, int col)
    {
        if (row < 0 || row > 3 || col < 0 || col > 3)
            return false;

        Die die = this.dice[row,col];

        if (die.marked)
            return false;
        else if (search[this.compareLength] != die.letter)
            return false;
        else if ((search.Length - 1) == this.compareLength)
            return true;
        else
        {
            die.marked = true;
            this.compareLength++;

            for (int i = -1; i <= 1; i++)
            {
                for (int j = -1; j <= 1; j++)
                {
                    if (!(i == 0 && j == 0))
                    {
                       if (this.stringFound(search, row + i, col + j))
                       {
                           die.marked = false;
                           return true;
                       }
                    }
                }
            }
        }

        die.marked = false;
        this.compareLength--;

        return false;
    }

    public string ToPattern()
    {
        StringBuilder sb = new StringBuilder();

        for (int row = 0; row < 4; row++)
            for (int col = 0; col < 4; col++)
                sb.Append(this.dice[row, col]);

        return sb.ToString();
    }

    public override String ToString()
    {
        StringBuilder sb = new StringBuilder();

        for (int row = 0; row < 4; row++)
        {
            for (int col = 0; col < 4; col++)
            {
                sb.Append(this.dice[row,col].ToString());
            }

            sb.AppendLine();
        }

        return sb.ToString();
    }

    public static IList<T> Shuffle<T>(IList<T> deck)
    {
        int N = deck.Count;
        for (int i = 0; i < N; ++i)
        {
            int r = i + (int)(random.Next(N - i));
            T t = deck[r];
            deck[r] = deck[i];
            deck[i] = t;
        }

        return deck;
    }
}

File: java/src/game/DiceTray.java
package game;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Random;

public class DiceTray
{
	private Die[][] dice;
	private int compareLength;
	
	public DiceTray()
	{
		Random random = new Random();
		
        ArrayList<String> godsDice = new ArrayList<String>();
        godsDice.ensureCapacity(16);
        godsDice.add("LRYTTE");
        godsDice.add("ANAEEG");
        godsDice.add("AFPKFS");
        godsDice.add("YLDEVR");
        godsDice.add("VTHRWE");
        godsDice.add("IDSYTT");
        godsDice.add("XLDERI");
        godsDice.add("ZNRNHL");
        godsDice.add("EGHWNE");
        godsDice.add("OATTOW");
        godsDice.add("HCPOAS");
        godsDice.add("NMIQHU");
        godsDice.add("SEOTIS");
        godsDice.add("MTOICU");
        godsDice.add("ENSIEU");
        godsDice.add("OBBAOJ");
        
        Collections.shuffle(godsDice);
        
        this.dice = new Die[4][4];
        for (int row = 0; row < 4; row++)
        {
            for (int col = 0; col < 4; col++)
            {
                this.dice[row][col] = new Die(godsDice.get(0).charAt(random.nextInt(6)));
                godsDice.remove(0);
            }
        }
	}
	
	public Die[][] getDice()
	{
		return this.dice;
	}
	
    public boolean stringFound(String search)
    {
        search = search.toUpperCase();
        this.compareLength = 0;
        char firstLetterOfSearch = search.charAt(0);

        for (int i = 0; i < 4; i++)
        {
            for (int j = 0; j < 4; j++)
            {
                if (firstLetterOfSearch == this.dice[i][j].getLetter())
                    if (this.stringFound(search, i, j))
                        return true;
            }
        }

        return false;
    }

    private boolean stringFound(String search, int row, int col)
    {
        if (row < 0 || row > 3 || col < 0 || col > 3)
            return false;

        Die die = this.dice[row][col];

        if (die.isMarked())
            return false;
        else if (search.charAt(this.compareLength) != die.getLetter())
            return false;
        else if ((search.length() - 1) == this.compareLength)
            return true;
        else
        {
            die.setMarked(true);
            this.compareLength++;

            for (int i = -1; i <= 1; i++)
            {
                for (int j = -1; j <= 1; j++)
                {
                    if (!(i == 0 && j == 0))
                    {
                       if (this.stringFound(search, row + i, col + j))
                       {
                           die.setMarked(false);
                           return true;
                       }
                    }
                }
            }
        }

        die.setMarked(false);
        this.compareLength--;

        return false;
    }

    public String toPattern()
    {
        StringBuilder sb = new StringBuilder();

        for (int row = 0; row < 4; row++)
            for (int col = 0; col < 4; col++)
                sb.append(this.dice[row][col]);

        return sb.toString().toLowerCase();
    }

    @Override
    public String toString()
    {
        StringBuilder sb = new StringBuilder();

        for (int row = 0; row < 4; row++)
        {
            for (int col = 0; col < 4; col++)
            {
                sb.append(this.dice[row][col].toString());
            }

            sb.append('\n');
        }

        return sb.toString();
    }
}
